### PR TITLE
feat: survey variables

### PR DESCRIPTION
**Pull request by pandeymangg**
## What does this PR do?

adds survey variables functionality
fixes https://github.com/formbricks/internal/issues/272

## How should this be tested?

- Open the survey editor
- Add some variables! (under questions tab)

### Required

- [x] Filled out the "How to test" section in this PR
- [x] Read [How we Code at Formbricks](<[https://github.com/formbricks/formbricks/blob/main/CONTRIBUTING.md](https://formbricks.com/docs/contributing/how-we-code)>)
- [x] Self-reviewed my own code
- [x] Commented on my code in hard-to-understand bits
- [x] Ran `pnpm build`
- [x] Checked for warnings, there are none
- [x] Removed all `console.logs`
- [x] Merged the latest changes from main onto my branch with `git pull origin main`
- [x] My changes don't cause any responsiveness issues
- [ ] First PR at Formbricks? [Please sign the CLA!](https://cla-assistant.io/formbricks/formbricks) Without it we wont be able to merge it 🙏

### Appreciated

- [ ] If a UI change was made: Added a screen recording or screenshots to this PR
- [ ] Updated the Formbricks Docs if changes were necessary

### CODE REVIEW

**Review by Dhruwang — CHANGES_REQUESTED** [review]
Hey @pandeymangg , feature works reallly well and code looks good too. Made few comments that can be helpful 😊

1. When variables card is selected in questionView, it shows an empty card in survey preview. We can make it similar to hidden fields, where preview does not changes 😊

Screenshot 2024-08-16 at 9 36 54 AM

2. We also need to add `variables`  to the list of forbidden ids. So that no questionId can be set to this value 

Screenshot 2024-08-16 at 11 43 51 AM

Can you please also fix the build and unit test 😊

File: apps/web/app/(app)/(survey-editor)/environments/[environmentId]/surveys/[surveyId]/edit/components/SurveyVariableCardsItem.tsx [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+import { Select, SelectContent, SelectItem, SelectTrigger, SelectValue } from "@formbricks/ui/Select";
+
+interface SurveyVariableCardsItemProps {
+  variable?: TSurveyVariable;
+  localSurvey: TSurvey;
+  setLocalSurvey: React.Dispatch<React.SetStateAction<TSurvey>>;
+  mode: "create" | "edit";
+}
+
+export const SurveyVariableCardsItem = ({
```
[the remark this answers is not in the document]
Comment (mattinannt) [reply apps/web/app/(app)/(survey-editor)/environments/[environmentId]/surveys/[surveyId]/edit/components/SurveyVariableCardsItem.tsx:22]: `SurveyVariablesCardItem` ;-)

**Review by mattinannt — CHANGES_REQUESTED** [review]
@pandeymangg Thanks for the updates :-) I like that you solved the open state of the variables card independently from the questionId workaround, but unfortunately since you only changed that for the variables tab the behaviour of the cards is now inconsistent.
The current behaviour was that only one card can be opened and the opened card gets closed when a new card is opened. Please make the following changes:
- build a global state for the currently opened card but make sure this is not directly tied to the `questionId`, but so generic that it works for the different card types and we don't need a workaround like the FORBIDDEN_IDS which solve multiple purposes.
- for now hide / comment out the `Variables` card for us to merge the PR with the structural changes only since the variables only make sense once the advanced logic is merged.

**Review by mattinannt — APPROVED** [review]
@pandeymangg thanks for making the changes :-) Works great!!! 🚀🔥

I commented out the variables card in the survey editor for now and we will show it again with the advanced logic merge 😊.

**Pull request merged**